### PR TITLE
Implemented mongodb storage

### PR DESCRIPTION
**Pull request by asimaranov**
Implemented the mongodb storage. Added some tests like in redis storage. Added motor dependency

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -49,6 +49,9 @@ Babel = { version = "^2.9.1", optional = true }
 aiohttp-socks = { version = "^0.5.5", optional = true }
 # Redis
 aioredis = { version = "^2.0.0", optional = true }
+# Mongodb
+motor = "^2.5.1"
```
Comment (uburuntu) [inline pyproject.toml:53]: Make this optional, please, as `aioredis` from previous lines

File: tests/conftest.py [hunk 2]
```diff
@@ -67,6 +68,19 @@ async def memory_storage():
         await storage.close()
 
 
+@pytest.fixture()
+@pytest.mark.mongo
+async def mongo_storage(redis_server):
+    if not redis_server:
+        pytest.skip("Mongo is not available here")
+    storage = MongoStorage.from_url(redis_server)
```
Comment (uburuntu) [inline tests/conftest.py:76]: And change the variable name, it's confusing currently :)

**Review by uburuntu — CHANGES_REQUESTED** [review]
Thanks for the PR! Overall lgtm.

I'm not familiar with MongoDB, so hope someone will check its usage.

File: aiogram/dispatcher/fsm/storage/mongo.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+from typing import Any, AsyncGenerator, Dict, Optional
+
+try:
+    import motor
+    from motor.motor_asyncio import AsyncIOMotorClient
+except ModuleNotFoundError as e:
+    import warnings
+
+    warnings.warn("Install motor with `pip install motor`")
+    raise e
```
Comment (Tishka17) [inline aiogram/dispatcher/fsm/storage/mongo.py:11]: Just `raise`

File: pyproject.toml [hunk 4]
```diff
@@ -50,7 +50,7 @@ aiohttp-socks = { version = "^0.5.5", optional = true }
 # Redis
 aioredis = { version = "^2.0.0", optional = true }
 # Mongodb
-motor = "^2.5.1"
+motor = { version = "^2.5.1", optional = true }
```
Comment (WhiteMemory99) [inline pyproject.toml:53]: It would be nice to add this dependency to the extras section below.

**Comment by evgfilim1** [comment]
Hello, @asimaranov, and thanks for contributing! If you've finished working on this feature, please reformat your code with `make reformat` so the tests can pass.

**Comment by asimaranov** [comment]
Why can it throw "E   ModuleNotFoundError: No module named 'motor'"?

**Comment by evgfilim1** [comment]
> Why can it throw "E ModuleNotFoundError: No module named 'motor'"?

@asimaranov, because it's not installed during testing. Look how Redis is tested in [`tests/`](https://github.com/aiogram/aiogram/tree/dev-3.x/tests) directory and [`.github/workflows/tests.yml`](https://github.com/aiogram/aiogram/blob/dev-3.x/.github/workflows/tests.yml) and set it up similarly.

File: poetry.lock [hunk 5]
```diff
@@ -568,6 +568,20 @@ category = "dev"
 optional = false
 python-versions = "*"
 
+[[package]]
+name = "motor"
+version = "2.5.1"
+description = "Non-blocking MongoDB driver for Tornado or asyncio"
+category = "main"
+optional = false
```
Comment (evgfilim1) [inline poetry.lock:576]: Please update lockfile also, this should become `true`
https://github.com/aiogram/aiogram/blob/6e39f9fada99906567d889c9618f65bc2f089c44/poetry.lock#L47

**Review by evgfilim1 — CHANGES_REQUESTED** [review]

**Comment by KobetsDev** [comment]
can you tell me why PR was abandoned?

**Comment by JrooTJunior** [comment]
> can you tell me why PR was abandoned?

The author of this PR is inactive and has not responded for a long time, but the code from this PR can be reused to implement Mongo storage.

**Pull request closed without merging**